compiler/test/runnable/testcmain.d

```diff
@@ -0,0 +1,3 @@
+module object;
+extern(D) void main() { }
+extern(C) void main() { }
```

Comment:
> This will fail on LDC which checks for duplicate main files. https://github.com/ldc-developers/ldc/blob/d4f2bed3a45688b32434d429e39c22dea5a9ffda/gen/functions.cpp#L550
> 

Doesn't look like it will fail to me, even the comment says.

"""
a D main implies a C main, so only check C mains with -betterC
"""

> Also this test doesn't seem to check anything, are we implicitly relying on the test infra to check that the return code is zero? If so could you add some brief comments to this as there is not link to a bug report.

I'm surprised you even think a runnable/compilable test exiting with a non-zero code should be considered success.

FYI a fail_compilation test should fail also when the exit code isn't 1 as well - for example, an exit code indicating a fault (illegal instruction, bus, floating point, or otherwise) is not a successful failure test.

Reply:
Hmm,
```d
extern(D) void main() { }
extern(C) void main() { }
``` does fail with 
```
/dlang/ldc-1.31.0/bin/../import/core/internal/entrypoint.d(39): Error: Function type does not match previously declared function with the same mangled name: `main`
/dlang/ldc-1.31.0/bin/../import/core/internal/entrypoint.d(39):        Previous IR type: i32 ()
/dlang/ldc-1.31.0/bin/../import/core/internal/entrypoint.d(39):        New IR type:      i32 (i32, i8**)
```
cc @kinke

Reply:
You shouldn't be getting that with `module object;` at the top.  Yes, I did notice that the behaviour of the front-end is to instantiate the `_d_cmain` template blindly if it finds it and D main exists.

Reply:
The frontend change is absolutely fine IMO, but the testcase is problematic. The `_d_cmain` template generates a conflicting C main (taking 2 args); with LDC, you get the error above, while with DMD, you apparently get a C main where *both* bodies are glued together (from [run.dlang.io](https://run.dlang.io/is/cIF7T7)):
```D
extern(C) void main() { assert(0); }
extern(D) void main() { assert(0); }
```
=>
```asm
.text.main	segment
	assume	CS:.text.main
main:
main:
		push	RBP
		mov	RBP,RSP
		mov	ESI,1
		lea	RDI,_TMP0@PC32[RIP]
		call	  _d_assertp@PLT32
		pop	RBP
		ret
		nop
		push	RBP
		mov	RBP,RSP
		sub	RSP,010h
		mov	-010h[RBP],EDI
		mov	-8[RBP],RSI
		mov	RDX,_Dmain@GOTPCREL[RIP]
		mov	RSI,-8[RBP]
		mov	EDI,-010h[RBP]
		call	  _d_run_main@PLT32
		leave
		ret
```

Reply:
Oh sorry, I see the `module object;` at the top now - yep, no more `_d_cmain` instantiation then. But the test is still pretty weak, obviously compiles fine with existing compilers. The `return 0` would probably only really be testable at the AST level; oh well.